src/detect.rs

```diff
@@ -18,3 +23,221 @@ pub enum DetectOutcome {
     Pass(BuildPlan),
     Fail,
 }
+
+pub struct DetectContextBuilder<P: Platform, BM: DeserializeOwned> {
+    app_dir: Option<PathBuf>,
+    buildpack_dir: Option<PathBuf>,
+    stack_id: String,
+    platform: Option<P>,
+    platform_dir: Option<PathBuf>,
+    buildpack_toml: Option<BuildpackToml<BM>>,
+    buildpack_toml_path: Option<PathBuf>,
+    buildpack_toml_string: Option<String>,
+}
+impl<P: Platform, BM: DeserializeOwned> DetectContextBuilder<P, BM> {
+    pub fn new(stack_id: impl AsRef<str>) -> Self {
+        DetectContextBuilder {
+            app_dir: None,
+            buildpack_dir: None,
+            buildpack_toml_path: None,
+            buildpack_toml_string: None,
+            platform_dir: None,
+            stack_id: String::from(stack_id.as_ref()),
+            platform: None,
+            buildpack_toml: None,
+        }
+    }
+
+    /// Used for testing purposes to set a minimal viable context
+    ///
+    ///```
+    /// use tempfile;
+    /// use libcnb::DetectContextBuilder;
+    ///
+    /// pub fn detect(context: libcnb::GenericDetectContext) {}
+    ///
+    /// let temp_dir = tempfile::tempdir().unwrap();
+    /// let context = DetectContextBuilder::new("heroku-20")
+    ///     .buildpack_toml_from_string(include_str!("../examples/example-02-ruby-sample/buildpack.toml"))
+    ///     .from_temp_dir(&temp_dir)
+    ///     .build::<std::io::Error>()
+    ///     .unwrap();
+    ///
+    /// detect(context);
+    ///```
+    pub fn from_temp_dir(mut self, temp_dir: &tempfile::TempDir) -> DetectContextBuilder<P, BM> {
+        let app_dir = temp_dir.path().join("app");
+        let platform_dir = temp_dir.path().join("platform");
+        let buildpack_dir = temp_dir.path().join("buildpack");
+
+        for dir in [&app_dir, &platform_dir, &buildpack_dir] {
+            std::fs::create_dir_all(dir).unwrap();
+        }
+
+        if self.app_dir.is_none() {
+            self.app_dir = Some(app_dir);
+        }
+
+        if self.buildpack_dir.is_none() {
+            self.buildpack_dir = Some(buildpack_dir);
+        }
+
+        if self.platform.is_none() {
+            self.platform_dir = Some(platform_dir);
+        }
+        self
+    }
+
+    pub fn buildpack_toml_from_string(
+        mut self,
+        body: impl AsRef<str>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml_string = Some(String::from(body.as_ref()));
+        self
+    }
+
+    pub fn buildpack_toml_from_file(
+        mut self,
+        path: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml_path = Some(path.as_ref().to_path_buf());
+        self
+    }
```

Comment:
I wonder if we could take an enum here and then match it on build:

```rust
enum BuildpackTomlArg {
  Path(PathBuf),
  Str(String)
}

pub fn buildpack_toml(input: BuildpackTomlArg) {
  // ...
}
```

Reply:
I like that.

Reply:
This would make for a more verbose API if we expose it:

```
    /// let temp_dir = tempfile::tempdir().unwrap();
    /// let context = DetectContextBuilder::new("heroku-20")
    ///     .buildpack_toml(BuildpackTomlArg::Str(include_str!("../examples/example-02-ruby-sample/buildpack.toml")))
    ///     .from_temp_dir(&temp_dir)
    ///     .build::<std::io::Error>()
    ///     .unwrap();
```

But if we don't expose it, and set it as a value internally it solves the problem of more than one source of truth, in this case the last one wins. Which might be confusing but seems fine. I tried it out and it does make some of the match stuff cleaner too reason about:

```
    fn set_blerg(mut self, platform: BuildpackToml<BM>) -> DetectContextBuilder<P, BM> {
        self.blerg = Some(BuildpackTomlArg::BuildpackToml(platform));
        self
    }

    fn set_blerg_string(mut self, string: impl AsRef<str>) -> DetectContextBuilder<P, BM> {
        self.blerg = Some(BuildpackTomlArg::Str(string.as_ref().to_string()));
        self
    }

    fn set_blerg_path(mut self, path: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
        self.blerg = Some(BuildpackTomlArg::Path(path.as_ref().to_path_buf()));
        self
    }
```

```
        if let Some(arg) = self.blerg {
            let foo = match arg {
                BuildpackTomlArg::BuildpackToml(buidpack_toml) => buildpack_toml,
                BuildpackTomlArg::Str(contents) => crate::toml_file::from_string(contents)
                    .map_err(Error::CannotReadBuildpackDescriptor)?,
                BuildpackTomlArg::Path(path) => crate::toml_file::read_toml_file(path)
                    .map_err(Error::CannotReadBuildpackDescriptor)?,
            };
        };
```

> Note: Using blerg here as a temp replacement.

Reply:
Or

```rs
let buildpack_toml = self
    .buildpack_toml
    .ok_or_else(|| {
        crate::error::Error::Other(
            "Platform is missing, call platform() or platform_from_path()".to_string(),
        )
    })
    .and_then(|arg| match arg {
        BuildpackTomlArg::BuildpackToml(a) => Ok(a),
        BuildpackTomlArg::Str(contents) => crate::toml_file::from_string(contents)
            .map_err(Error::CannotReadBuildpackDescriptor),
        BuildpackTomlArg::Path(path) => crate::toml_file::read_toml_file(path)
            .map_err(Error::CannotReadBuildpackDescriptor),
    })?;
```

Reply:
I really like the result. I pushed up my code after porting everything to an enum.